### ISSUE TITLE
We should make test phase another file, so that access the test set only if you need to submit a paper.

### ISSUE THREAD

**Issue by LuoXin-s** [ashleve/lightning-hydra-template#103]

**Comment by hobogalaxy** [comment]
@LuoXin-s 
Might be a good idea but we would need to investigate if it doesn't make a template too complicated or too specific for given task.

Could you elaborate on why is accessing the test set important only when submitting the paper? I'm not sure what you mean here...

I guess we could create `test.py` file next to `train.py` file, the question is how should the process look like:
1. One option is that user would need to execute testing file separately, this would require to give checkpoint path to trained model as an argument every time, which is not very convenient
2. Another option is to separete logic but still keep it as a single pipeline - train.py would still always execute testing after training ends, but testing logic would simply be moved to another file.

Currently the whole testing is simply calling `trainer.test()`. What are the advantages of moving this to another file? Or maybe you would like to create a completely new model and datamodel for testing only? Or maybe do you mean to make it more like a custom inference, similar to [inference_example.py](https://github.com/hobogalaxy/lightning-hydra-template/blob/main/src/utils/inference_example.py)?

I think one disadvantage of testing separately is that it's hard to keep track of which training logs are connected to which testing logs, even when using experiment loggers like wandb.

**Comment by LuoXin-s** [comment]
@hobogalaxy 
I think the most important things about **test set** is how does **test set** actually mean. The train/validation/test split it to enable train/validation pipeline, one train model on train set, and evaluate one's model on validation set,  and report scores on test set only when they satisfied their model and ready to put their work to production environment or paper. Actually if one can access test set in early phase one may to adjust their model according to results on test set, close the advantages of generability that the train/validation/test split brought.

And in my practice I used seperate `test.py`.

**Comment by hobogalaxy** [comment]
I get what you mean, we shouldn't optimize for test set.
What I usually do is logging both validation and test results for all experiments, but optimize only for validation dataset. 
So for example, I'm doing hyperparameter search, checkpointing and early stopping based on validation accuracy only and I choose the final model ignoring test results. The problem is, users can easily confuse it.

Also, I still sometimes like to have test results for all experiments simply to see how much it varies on average in comparison to validation results.

Encouraging separation of testing from training is definietely of value. I will investigate how it can be incorporated in a clean way and what should the workflow look like.

**Comment by lollows** [comment]
> What I usually do is logging both validation and test results for all experiments, but optimize only for validation datasets.

I agree with you. This is the right way to do the training. Because the validation set is made to guarantee the training quality. If the results of test phase has a large difference with validation phase, the problem is the result of your incorrect choice of training datasets, e.g., training on **cifar10** but testing on **minst**.


> maybe do you mean to make it more like a custom inference, similar to [inference_example.py](https://github.com/hobogalaxy/lightning-hydra-template/blob/main/src/utils/inference_example.py)?

I think the actual thing you want is a quick reference @LuoXin-s (this is slightly more complicated. You need to pruning or do some other things to accelerate you speed)


> Actually if one can access test set in early phase one may to adjust their model according to results on test set

In fact,  we do not have access to a test set in many cases (:worried: e.g., kaggle competition)
